design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,164 @@
+Declaring the field as dynamic will be done by adding a `cpu` entry to a new section under the VM spec, called `liveUpdateFeatures`.
+Under the `cpu` entry, it will be possible to define a maximum number of sockets. That number is needed by LibVirt and will default to 4 times the initial number of sockets if not set. That default value will be configurable in the KubeVirt CR.
+On VM startup, that value will translate to a new `maxSockets` entry under `spec.domain.cpu` in the VMI object. However, all VMI CPU fields will stay immutable, and CPU sockets will only be adjustable on the VM object.
+
+Increasing the number of CPU sockets will not only add CPUs to the guest but also increase the CPU resources available to the virt-launcher pod, via a migration (more on that in the Implementation section).
+The way LibVirt handles CPU hotplug is by exposing the maximum number of CPUs to the guest and turning off the unused CPUs. In this document, when we refer to "adding" CPUs to the guest, we're really just turning them on. This is why we need to define a maximum number of sockets in advance.
+
+## Drawbacks / Limitations
+- Since CPU hot-(un)plug involves a live migration, VMs will have to be `LiveMigratable` to enable CPU hotplug
+- Each disabled vCPU, so `(maxSockets - sockets) * cores * thread`, consumes 8MiB of overhead memory
+- This feature will be incompatible with CPU requests/limits, at least initially, since those values need to dynamically scale according to the current number of enabled CPUs
+
+## API Examples
+
+### VM Spec API
+
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-cirros
+  name: vm-cirros
+spec:
+  running: false
+  liveUpdateFeatures:
+    cpu
+      maximumSockets: 4
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      domain:
+        cpu:
+          sockets: 2
+          cores: 4
+          threads: 2
+```
+In this example, the user will be able to double the guest CPU resources by increasing the value of `sockets`.
+
+### VMI Spec API
+
+The VM definition above would translate to the following VMI:
+
+
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  labels:
+    special: vm-cirros
+  name: vm-cirros
+spec:
+  domain:
+    cpu:
+      maxSockets: 4
+      sockets: 2
+      cores: 4
+      threads: 2
+```
+
+### Status API
+
+The VMI status must indicate the current number of CPU sockets. The value needs to be propagated back to the VM status to allows the user to have an indication of when their changes applied to the running VMI.
+
+The user should be able to inspect the VM.Status value below to determine the state of their changes. This is similar in concept to someone inspecting a Deployments status to determine the the number of pods that are available.
+
+```
+// The number of sockets the vm has currently active.
+vmi.Status.Resources.Sockets.Ready
+vm.Status.Resources.Sockets.Ready
+```
```

Comment:
Would it make sense to simply keep the `CurrentCPUTopology` as a whole in the status? I know that only sockets change, but that would show the actual current topology and no one would be confused.
I just had something like that in mind: ```vmiCopy.Status.CurrentCPUTopology = vmi.Spec.Domain.CPU```